# 📧 SMS Spam Classification using Naive Bayes

## 🚀 Introduction

This project focuses on building a robust SMS spam classification system using machine learning techniques. Spam messages are a significant nuisance and security threat, making an automated detection system highly valuable. We will explore data preprocessing, model training, and evaluation to accurately distinguish between legitimate (ham) and spam messages.

## 🎯 Problem Statement

The goal is to classify incoming SMS messages as either 'spam' or 'not spam' (ham). This binary classification problem requires transforming unstructured text data into a suitable numerical format for a machine learning model to learn and make predictions. A high-accuracy model is crucial to minimize false positives (marking legitimate messages as spam) and false negatives (allowing spam messages to pass through).

## ✨ Objectives

*   **Data Preprocessing:** Clean and prepare the SMS text data for machine learning.
*   **Feature Extraction:** Convert text messages into numerical features using appropriate techniques.
*   **Model Building:** Develop a classification model capable of distinguishing spam from non-spam messages.
*   **Model Evaluation:** Assess the model's performance using standard classification metrics.
*   **Prediction System:** Create a simple interface to classify new, unseen SMS messages.

## 📊 Dataset Description

The dataset used for this project is a collection of SMS messages, each labeled as either 'ham' (legitimate) or 'spam'). It typically contains two main columns:

*   `Category`: The label indicating whether the message is 'ham' or 'spam'.
*   `Masseges`: The actual text content of the SMS message.

This dataset provides a good foundation for training and evaluating a text classification model.

In [ ]:
# 📚 Import Libraries
# This cell imports essential Python libraries for data manipulation and numerical operations.
# `pandas` is used for handling and analyzing tabular data (DataFrames).
# `numpy` provides support for large, multi-dimensional arrays and mathematical functions.
import numpy as np
import pandas as pd

## 📚 Import Libraries

This section is dedicated to importing all the necessary Python libraries. We use `pandas` for efficient data handling and `numpy` for numerical operations, which are fundamental for almost any data science project.

## ⚙️ Project Workflow

The project follows a standard machine learning pipeline:

1.  **Import Libraries**: Essential tools for data manipulation and model building.
2.  **Data Loading**: Ingesting the SMS message dataset.
3.  **Exploratory Data Analysis (EDA)**: Understanding the dataset's structure and characteristics.
4.  **Data Cleaning & Preprocessing**: Cleaning and preparing text data for model input.
5.  **Feature Engineering**: Converting text into numerical features suitable for machine learning algorithms.
6.  **Model Building & Training**: Selecting and training a classification model.
7.  **Model Evaluation**: Assessing the model's performance using various metrics.
8.  **Prediction**: Demonstrating the model's ability to classify new messages.
9.  **Conclusion & Future Work**: Summarizing findings and outlining potential improvements.

In [ ]:
# 📂 Load Dataset
# We load the SMS spam dataset from a CSV file into a pandas DataFrame.
# This step is crucial as it brings our raw data into a structured format for further processing.
df = pd.read_csv("/content/drive/MyDrive/Text Classification/spam mail.csv")

## 📂 Data Loading

Loading the dataset is the first step to bring our raw information into the analysis environment. We read the SMS spam dataset from a CSV file into a pandas DataFrame, which provides a structured and convenient way to work with the data.

In [ ]:
# 🔍 Initial Data Inspection
# To get a quick overview of the dataset structure and content, we display the first 5 rows.
# This helps in understanding the columns, data types, and initial patterns.
df.head()

,Category,Masseges
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


### ✨ Observations from Initial Data Inspection

The `df.head()` output provides a quick glance at the dataset:
*   **Columns:** We can observe two main columns: `Category` (our target variable, indicating 'ham' or 'spam') and `Masseges` (the text content of the SMS).
*   **Data Structure:** The data is tabular, with each row representing an SMS message and its corresponding category.
*   **Initial Categories:** The `Category` column initially contains 'ham' and 'spam' labels.

## 🔍 Exploratory Data Analysis (EDA)

Exploratory Data Analysis (EDA) is crucial for understanding the characteristics of our dataset. We'll examine its structure, check for imbalances, and get a general sense of the data before diving into preprocessing and modeling.

### ✨ Observations from Initial Data Inspection

The `df.head()` output provides a quick glance at the dataset:
*   **Columns:** We can observe two main columns: `Category` (our target variable, indicating 'ham' or 'spam') and `Masseges` (the text content of the SMS).
*   **Data Structure:** The data is tabular, with each row representing an SMS message and its corresponding category.
*   **Initial Categories:** The `Category` column initially contains 'ham' and 'spam' labels.

In [ ]:
# 🧹 Data Cleaning: Category Renaming & Distribution
# For clarity and consistency, we rename the 'ham' category to 'Not Spam'.
# Subsequently, we print the value counts of the 'Category' column to understand the class distribution (spam vs. not spam).
# Displaying the head again confirms the renaming.
df['Category'] = df['Category'].replace({'ham': 'Not Spam'})
print("Class Distribution after renaming 'ham' to 'Not Spam':")
print(df['Category'].value_counts())
print("\nDataFrame Head after category renaming:")
df.head()

Class Distribution after renaming 'ham' to 'Not Spam':
Category
Not Spam    4825
spam         747
Name: count, dtype: int64

DataFrame Head after category renaming:


,Category,Masseges
0,Not Spam,"Go until jurong point, crazy.. Available only ..."
1,Not Spam,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,Not Spam,U dun say so early hor... U c already then say...
4,Not Spam,"Nah I don't think he goes to usf, he lives aro..."


## 🧹 Data Cleaning & Preprocessing

This phase focuses on preparing the text data for machine learning. Text data often contains noise and needs to be transformed into a numerical format that models can understand. For this project, the primary preprocessing steps involve standardizing categories and vectorizing text.

### 📊 Class Distribution Analysis

After renaming 'ham' to 'Not Spam', the class distribution reveals a significant imbalance:
*   **'Not Spam' messages:** 4825
*   **'Spam' messages:** 747

This imbalance (approximately 6.5 times more 'Not Spam' than 'Spam' messages) is typical in spam detection datasets. It's an important observation because it can affect model training and evaluation. A model might achieve high overall accuracy by simply predicting 'Not Spam' most of the time. Therefore, metrics like precision, recall, and F1-score will be crucial in evaluating performance, especially for the minority 'spam' class.

## 🧹 Data Cleaning & Preprocessing

This phase focuses on preparing the text data for machine learning. Text data often contains noise and needs to be transformed into a numerical format that models can understand. For this project, the primary preprocessing steps involve standardizing categories and vectorizing text.

### 📊 Class Distribution Analysis

After renaming 'ham' to 'Not Spam', the class distribution reveals a significant imbalance:
*   **'Not Spam' messages:** 4825
*   **'Spam' messages:** 747

This imbalance (approximately 6.5 times more 'Not Spam' than 'Spam' messages) is typical in spam detection datasets. It's an important observation because it can affect model training and evaluation. A model might achieve high overall accuracy by simply predicting 'Not Spam' most of the time. Therefore, metrics like precision, recall, and F1-score will be crucial in evaluating performance, especially for the minority 'spam' class.

In [ ]:
# 📊 Data Splitting: Training and Testing Sets
# We split the dataset into training and testing sets using `train_test_split`.
# This is vital to evaluate the model's performance on unseen data, preventing overfitting.
# `x` (features) is the 'Masseges' column, and `y` (target) is the 'Category' column.
# A `test_size` of 20% means 80% of data is for training and 20% for testing, with `random_state` ensuring reproducibility.
from sklearn.model_selection import train_test_split
x = df["Masseges"]
y = df["Category"]
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=2)

## ⚙️ Feature Engineering & Text Preprocessing

For text classification, raw text needs to be converted into numerical features that machine learning models can process. This involves techniques like tokenization and vectorization. Here, `CountVectorizer` is employed for this purpose.

### ✨ Observations from Data Splitting

After splitting, we have:
*   `x_train`: Features (SMS messages) for training the model.
*   `x_test`: Features (SMS messages) for evaluating the model.
*   `y_train`: Corresponding labels (Category) for training.
*   `y_test`: Corresponding labels (Category) for evaluation.

The sizes of these sets are determined by the `test_size` (20%) and the original dataset size, ensuring a clear separation for unbiased evaluation.

### 🔪 Data Splitting Rationale

Splitting the data into training and testing sets is a fundamental practice in machine learning to prevent overfitting and ensure the model's ability to generalize to new, unseen data. The training set is used to teach the model patterns, while the test set provides an unbiased evaluation of its performance. A `test_size` of 0.2 means 20% of the data is reserved for testing, which is a common split ratio. The `random_state` parameter ensures that the split is reproducible, yielding the same training and test sets each time the code is run.

## ⚙️ Feature Engineering & Text Preprocessing

For text classification, raw text needs to be converted into numerical features that machine learning models can process. This involves techniques like tokenization and vectorization. Here, `CountVectorizer` is employed for this purpose.

### ✨ Observations from Data Splitting

After splitting, we have:
*   `x_train`: Features (SMS messages) for training the model.
*   `x_test`: Features (SMS messages) for evaluating the model.
*   `y_train`: Corresponding labels (Category) for training.
*   `y_test`: Corresponding labels (Category) for evaluation.

The sizes of these sets are determined by the `test_size` (20%) and the original dataset size, ensuring a clear separation for unbiased evaluation.

### 🔪 Data Splitting Rationale

Splitting the data into training and testing sets is a fundamental practice in machine learning to prevent overfitting and ensure the model's ability to generalize to new, unseen data. The training set is used to teach the model patterns, while the test set provides an unbiased evaluation of its performance. A `test_size` of 0.2 means 20% of the data is reserved for testing, which is a common split ratio. The `random_state` parameter ensures that the split is reproducible, yielding the same training and test sets each time the code is run.

In [ ]:
# ⚙️ Model Building: Text Vectorization and Naive Bayes Training
# We use `CountVectorizer` to convert text messages into a matrix of token counts, a common step in NLP for numerical representation.
# `fit_transform` is applied to the training data (`x_train`) to learn the vocabulary and transform it, while `transform` is used on `x_test` to apply the learned vocabulary.
# A `MultinomialNB` (Naive Bayes) model, suitable for text classification with discrete features (like word counts), is then initialized and trained (`fit`) on the vectorized training data.
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import CountVectorizer

# Initialize CountVectorizer to convert text into a matrix of token counts
vectorizer = CountVectorizer()

# Fit the vectorizer on the training data and transform it
x_train_v = vectorizer.fit_transform(x_train)

# Transform the test data using the *same* fitted vectorizer
x_test_v = vectorizer.transform(x_test)

# Initialize and train a Multinomial Naive Bayes model
# Multinomial Naive Bayes is a probabilistic classifier suitable for classification with discrete features (e.g., word counts for text classification).
model = MultinomialNB()
model.fit(x_train_v,y_train)

MultinomialNB()

### ✅ Model Training Confirmation

The output `MultinomialNB()` confirms that the Multinomial Naive Bayes model has been successfully initialized and trained on the vectorized training data (`x_train_v`) and their corresponding labels (`y_train`). The model is now ready to make predictions on unseen data.

### 🛠️ CountVectorizer: Choice, Advantages, and Limitations

**Choice Rationale:** `CountVectorizer` is chosen for its simplicity and effectiveness in converting a collection of text documents to a matrix of token counts, a common step in NLP for numerical representation. It's a foundational step for many text classification tasks and works well with Naive Bayes models.

**Advantages:**
*   **Simplicity:** Easy to understand and implement.
*   **Efficiency:** Relatively fast for large datasets.
*   **Interpretability:** The resulting features (word counts) are directly interpretable, indicating the frequency of words.
*   **Baseline Performance:** Often provides a strong baseline for text classification tasks.

**Limitations:**
*   **Sparsity:** Creates very sparse matrices, especially with a large vocabulary, which can be memory-intensive.
*   **No Semantic Information:** Treats each word as an independent feature, ignoring semantic relationships, context, or word order.
*   **Out-of-Vocabulary (OOV) words:** Cannot handle words not seen during training, assigning them a count of zero in new data.
*   **Stop Words and Inflection:** By default, it may not handle common words (stop words) or word variations (e.g., 'run', 'running') optimally without further configuration.

### 🧠 Multinomial Naive Bayes: Choice, Advantages, and Limitations

**Choice Rationale:** Multinomial Naive Bayes is a probabilistic classifier particularly well-suited for text classification problems where features represent counts (like word frequencies from `CountVectorizer`). It operates on the principle of conditional independence between features, given the class.

**Advantages:**
*   **Simplicity and Speed:** Easy to implement and computationally efficient, making it fast for both training and prediction.
*   **Scalability:** Performs well with large datasets and high-dimensional feature spaces, which is common in text data.
*   **Good Baseline:** Often provides competitive baseline performance for text classification tasks.
*   **Less Data Requirement:** Can perform well even with relatively small training datasets compared to more complex models.

**Limitations:**
*   **Feature Independence Assumption:** The core assumption that features are conditionally independent is often violated in real-world text data (e.g., 'New' and 'York' are not independent). Despite this, it often performs surprisingly well.
*   **Zero-Frequency Problem:** If a word from the test data was not present in the training data, its probability will be zero, leading to an incorrect prediction. This is usually mitigated by techniques like Laplace smoothing (often built into implementations).

**Expected Impact on Performance:** The combination of `CountVectorizer` and `Multinomial Naive Bayes` is expected to provide a robust and efficient solution for spam detection. While `CountVectorizer` captures word frequencies, `Multinomial Naive Bayes` leverages these frequencies to probabilistically classify messages. Given the nature of spam (frequent use of specific keywords), this approach should yield high accuracy, especially in distinguishing between spam and non-spam messages based on word occurrences.

## 🧠 Model Building & Training

This section details the selection, configuration, and training of our machine learning model. For text classification, a common and effective approach involves converting text into numerical features, followed by applying a suitable classification algorithm. We employ `CountVectorizer` for feature extraction and `Multinomial Naive Bayes` for classification.

## 🧠 Model Building & Training

This section details the selection, configuration, and training of our machine learning model. For text classification, a common and effective approach involves converting text into numerical features, followed by applying a suitable classification algorithm. We employ `CountVectorizer` for feature extraction and `Multinomial Naive Bayes` for classification.

### ✅ Model Training Confirmation

The output `MultinomialNB()` confirms that the Multinomial Naive Bayes model has been successfully initialized and trained on the vectorized training data (`x_train_v`) and their corresponding labels (`y_train`). The model is now ready to make predictions on unseen data.

### 🛠️ CountVectorizer: Choice, Advantages, and Limitations

**Choice Rationale:** `CountVectorizer` is chosen for its simplicity and effectiveness in converting a collection of text documents to a matrix of token counts. It's a foundational step for many text classification tasks and works well with Naive Bayes models.

**Advantages:**
*   **Simplicity:** Easy to understand and implement.
*   **Efficiency:** Relatively fast for large datasets.
*   **Interpretability:** The resulting features (word counts) are directly interpretable, indicating the frequency of words.
*   **Baseline Performance:** Often provides a strong baseline for text classification tasks.

**Limitations:**
*   **Sparsity:** Creates very sparse matrices, especially with a large vocabulary, which can be memory-intensive.
*   **No Semantic Information:** Treats each word as an independent feature, ignoring semantic relationships, context, or word order.
*   **Out-of-Vocabulary (OOV) words:** Cannot handle words not seen during training, assigning them a count of zero in new data.
*   **Stop Words and Inflection:** By default, it may not handle common words (stop words) or word variations (e.g., 'run', 'running') optimally without further configuration.

### 🧠 Multinomial Naive Bayes: Choice, Advantages, and Limitations

**Choice Rationale:** Multinomial Naive Bayes is a probabilistic classifier particularly well-suited for text classification problems where features represent counts (like word frequencies from `CountVectorizer`). It operates on the principle of conditional independence between features, given the class.

**Advantages:**
*   **Simplicity and Speed:** Easy to implement and computationally efficient, making it fast for both training and prediction.
*   **Scalability:** Performs well with large datasets and high-dimensional feature spaces, which is common in text data.
*   **Good Baseline:** Often provides competitive baseline performance for text classification tasks.
*   **Less Data Requirement:** Can perform well even with relatively small training datasets compared to more complex models.

**Limitations:**
*   **Feature Independence Assumption:** The core assumption that features are conditionally independent is often violated in real-world text data (e.g., 'New' and 'York' are not independent). Despite this, it often performs surprisingly well.
*   **Zero-Frequency Problem:** If a word from the test data was not present in the training data, its probability will be zero, leading to an incorrect prediction. This is usually mitigated by techniques like Laplace smoothing (often built into implementations).

**Expected Impact on Performance:** The combination of `CountVectorizer` and `Multinomial Naive Bayes` is expected to provide a robust and efficient solution for spam detection. While `CountVectorizer` captures word frequencies, `Multinomial Naive Bayes` leverages these frequencies to probabilistically classify messages. Given the nature of spam (frequent use of specific keywords), this approach should yield high accuracy, especially in distinguishing between spam and non-spam messages based on word occurrences.

In [ ]:
# ➡️ Make Predictions
# After training the model, we use it to predict the categories (spam or not spam) for the vectorized test messages (`x_test_v`).
# These predictions (`y_pred`) will be compared against the actual labels (`y_test`) for evaluation.
y_pred = model.predict(x_test_v)

## 📈 Model Evaluation

Evaluating the model's performance is critical to understand its effectiveness and identify areas for improvement. We will use several key metrics: Accuracy, Confusion Matrix, and Classification Report (Precision, Recall, F1-Score). These metrics provide a comprehensive view of how well our model performs on unseen data, especially considering the class imbalance present in the dataset.

### ➡️ Prediction Outcome

The `y_pred` variable now holds the model's predicted labels (either 'spam' or 'Not Spam') for each message in the test set. These predictions are the direct output of our trained `MultinomialNB` model when applied to the vectorized test features. The next step is to compare these predictions with the actual labels (`y_test`) to evaluate the model's performance comprehensively.

## 📈 Model Evaluation

Evaluating the model's performance is critical to understand its effectiveness and identify areas for improvement. We will use several key metrics: Accuracy, Confusion Matrix, and Classification Report (Precision, Recall, F1-Score). These metrics provide a comprehensive view of how well our model performs on unseen data, especially considering the class imbalance present in the dataset.

### ➡️ Prediction Outcome

The `y_pred` variable now holds the model's predicted labels (either 'spam' or 'Not Spam') for each message in the test set. These predictions are the direct output of our trained `MultinomialNB` model when applied to the vectorized test features. The next step is to compare these predictions with the actual labels (`y_test`) to evaluate the model's performance comprehensively.

In [ ]:
# 📈 Model Evaluation
# This cell calculates and displays key classification metrics: accuracy, confusion matrix, and a classification report.
# These metrics are crucial for understanding how well our model performs on unseen data.
# `accuracy_score` gives the overall correctness.
# `confusion_matrix` shows the counts of true positives, true negatives, false positives, and false negatives.
# `classification_report` provides precision, recall, and F1-score for each class, offering a more detailed performance breakdown.
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

print("\n--- Model Evaluation Results ---")
print(f"Accuracy Score: {accuracy_score(y_pred,y_test):.4f}")
print("\nConfusion Matrix:")
print(confusion_matrix(y_pred,y_test))
print("\nClassification Report:")
print(classification_report(y_pred,y_test))


--- Model Evaluation Results ---
Accuracy Score: 0.9803

Confusion Matrix:
[[954  19]
 [  3 139]]

Classification Report:
              precision    recall  f1-score   support

    Not Spam       1.00      0.98      0.99       973
        spam       0.88      0.98      0.93       142

    accuracy                           0.98      1115
   macro avg       0.94      0.98      0.96      1115
weighted avg       0.98      0.98      0.98      1115



## ✅ Conclusion

This project successfully developed and evaluated an SMS spam classification model using a Multinomial Naive Bayes algorithm. Through careful data loading, cleaning, text vectorization with `CountVectorizer`, and model training, we achieved a highly effective spam filter.

**Key Achievements:**
*   The model demonstrated strong performance with an overall accuracy of **98.03%**.
*   Crucially, it achieved a high **recall of 98% for the 'spam' class**, meaning it successfully identifies nearly all actual spam messages. This minimizes the critical 'false negatives' (spam messages that slip through).
*   The **precision for 'Not Spam' was 100%**, indicating that no legitimate messages were incorrectly flagged as spam (false positives).
*   The interactive prediction demo showcases the practical applicability of the trained model.

**Challenges and Lessons Learned:**
*   Handling class imbalance (more 'Not Spam' messages than 'Spam') required careful interpretation of metrics beyond just accuracy. The high recall for spam shows the model's robustness despite this imbalance.
*   The simplicity of `CountVectorizer` and `Multinomial Naive Bayes` provides an excellent balance of performance and computational efficiency for this specific problem.

**Practical Applications:**
This model can be integrated into messaging applications to automatically filter out unwanted spam, significantly enhancing user experience and reducing exposure to phishing attempts or unsolicited advertisements.

## 📊 Results Summary

The model's performance on the test set is summarized by the following metrics:

*   **Accuracy Score: 0.9803 (98.03%)**
    *   **Interpretation:** This indicates that approximately 98% of the messages in the test set were correctly classified by the model. While high, accuracy alone can be misleading in imbalanced datasets.

*   **Confusion Matrix:**
    ```
    [[954  19]
     [  3 139]]
    ```
    *   **True Negatives (TN = 954):** The model correctly predicted 954 'Not Spam' messages.
    *   **False Positives (FP = 19):** The model incorrectly predicted 19 'Not Spam' messages as 'spam'. These are Type I errors.
    *   **False Negatives (FN = 3):** The model incorrectly predicted 3 'spam' messages as 'Not Spam'. These are Type II errors, which are often more critical in spam detection as actual spam passes through.
    *   **True Positives (TP = 139):** The model correctly predicted 139 'spam' messages.

*   **Classification Report:**
    ```
                  precision    recall  f1-score   support

        Not Spam       1.00      0.98      0.99       973
            spam       0.88      0.98      0.93       142

        accuracy                           0.98      1115
       macro avg       0.94      0.98      0.96      1115
    weighted avg       0.98      0.98      0.98      1115
    ```

    *   **'Not Spam' Class (Majority Class):**
        *   **Precision (1.00):** Out of all messages predicted as 'Not Spam', 100% were actually 'Not Spam'. This is excellent, meaning very few legitimate messages are falsely flagged as spam.
        *   **Recall (0.98):** Out of all actual 'Not Spam' messages, the model correctly identified 98% of them. The 2% it missed were likely classified as spam (false positives).
        *   **F1-Score (0.99):** A high F1-score indicates a good balance between precision and recall for the 'Not Spam' class.

    *   **'Spam' Class (Minority Class):**
        *   **Precision (0.88):** Out of all messages predicted as 'spam', 88% were actually 'spam'. This means some messages predicted as spam were actually 'Not Spam' (false positives).
        *   **Recall (0.98):** Out of all actual 'spam' messages, the model correctly identified 98% of them. This is a crucial metric, as it shows the model is highly effective at catching spam.
        *   **F1-Score (0.93):** A strong F1-score for the spam class demonstrates good performance in identifying spam messages while keeping false positives at an acceptable level.

    *   **Overall:**
        *   The model exhibits a strong ability to detect spam, with a very high recall for the 'spam' class and excellent precision for the 'Not Spam' class. The low number of false negatives (only 3 spam messages missed) is particularly important for a spam filter. The accuracy is also very good, reflecting the overall correctness of the predictions.

## ✅ Conclusion

This project successfully developed and evaluated an SMS spam classification model using a Multinomial Naive Bayes algorithm. Through careful data loading, cleaning, text vectorization with `CountVectorizer`, and model training, we achieved a highly effective spam filter.

**Key Achievements:**
*   The model demonstrated strong performance with an overall accuracy of **98.03%**.
*   Crucially, it achieved a high **recall of 98% for the 'spam' class**, meaning it successfully identifies nearly all actual spam messages. This minimizes the critical 'false negatives' (spam messages that slip through).
*   The **precision for 'Not Spam' was 100%**, indicating that no legitimate messages were incorrectly flagged as spam (false positives).
*   The interactive prediction demo showcases the practical applicability of the trained model.

**Challenges and Lessons Learned:**
*   Handling class imbalance (more 'Not Spam' messages than 'Spam') required careful interpretation of metrics beyond just accuracy. The high recall for spam shows the model's robustness despite this imbalance.
*   The simplicity of `CountVectorizer` and `Multinomial Naive Bayes` provides an excellent balance of performance and computational efficiency for this specific problem.

**Practical Applications:**
This model can be integrated into messaging applications to automatically filter out unwanted spam, significantly enhancing user experience and reducing exposure to phishing attempts or unsolicited advertisements.

## 📊 Results Summary

The model's performance on the test set is summarized by the following metrics:

*   **Accuracy Score: 0.9803 (98.03%)**
    *   **Interpretation:** This indicates that approximately 98% of the messages in the test set were correctly classified by the model. While high, accuracy alone can be misleading in imbalanced datasets.

*   **Confusion Matrix:**
    ```
    [[954  19]
     [  3 139]]
    ```
    *   **True Negatives (TN = 954):** The model correctly predicted 954 'Not Spam' messages.
    *   **False Positives (FP = 19):** The model incorrectly predicted 19 'Not Spam' messages as 'spam'. These are Type I errors.
    *   **False Negatives (FN = 3):** The model incorrectly predicted 3 'spam' messages as 'Not Spam'. These are Type II errors, which are often more critical in spam detection as actual spam passes through.
    *   **True Positives (TP = 139):** The model correctly predicted 139 'spam' messages.

*   **Classification Report:**
    ```
                  precision    recall  f1-score   support

        Not Spam       1.00      0.98      0.99       973
            spam       0.88      0.98      0.93       142

        accuracy                           0.98      1115
       macro avg       0.94      0.98      0.96      1115
    weighted avg       0.98      0.98      0.98      1115
    ```

    *   **'Not Spam' Class (Majority Class):**
        *   **Precision (1.00):** Out of all messages predicted as 'Not Spam', 100% were actually 'Not Spam'. This is excellent, meaning very few legitimate messages are falsely flagged as spam.
        *   **Recall (0.98):** Out of all actual 'Not Spam' messages, the model correctly identified 98% of them. The 2% it missed were likely classified as spam (false positives).
        *   **F1-Score (0.99):** A high F1-score indicates a good balance between precision and recall for the 'Not Spam' class.

    *   **'Spam' Class (Minority Class):**
        *   **Precision (0.88):** Out of all messages predicted as 'spam', 88% were actually 'spam'. This means some messages predicted as spam were actually 'Not Spam' (false positives).
        *   **Recall (0.98):** Out of all actual 'spam' messages, the model correctly identified 98% of them. This is a crucial metric, as it shows the model is highly effective at catching spam.
        *   **F1-Score (0.93):** A strong F1-score for the spam class demonstrates good performance in identifying spam messages while keeping false positives at an acceptable level.

    *   **Overall:**
        *   The model exhibits a strong ability to detect spam, with a very high recall for the 'spam' class and excellent precision for the 'Not Spam' class. The low number of false negatives (only 3 spam messages missed) is particularly important for a spam filter. The accuracy is also very good, reflecting the overall correctness of the predictions.

In [ ]:
# 🚀 Real-time Prediction Demo
# This interactive cell allows a user to input a custom message and receive an immediate spam classification.
# The input message is first vectorized using the pre-fitted `CountVectorizer`, ensuring consistency with the training data.
# The trained `MultinomialNB` model then predicts its category (Spam or Not Spam), demonstrating the model's practical application.

user_input = input("Enter a message to classify: ")

# Vectorize the user's input using the *same* CountVectorizer used for training
user_input_v = vectorizer.transform([user_input])

# Make a prediction using the trained model
prediction = model.predict(user_input_v)

print(f"\nThe message is classified as: {prediction[0]}")

Enter a message to classify: hey how are you

The message is classified as: Not Spam


## 🔮 Future Improvements

To further enhance this SMS spam classification project and build upon the current success, several avenues for future work can be explored:

1.  **Hyperparameter Tuning:** Systematically optimize the parameters of `CountVectorizer` (e.g., `ngram_range`, `max_features`, `min_df`) and `MultinomialNB` (e.g., `alpha` for Laplace smoothing) using techniques like GridSearchCV or RandomizedSearchCV.

2.  **Advanced Text Preprocessing:** Incorporate more sophisticated text cleaning steps:
    *   **Stop Word Removal:** Eliminate common words (e.g., 'the', 'a', 'is') that often carry little semantic value.
    *   **Lemmatization/Stemming:** Reduce words to their root form (e.g., 'running' -> 'run') to minimize vocabulary size and treat variations as the same word.
    *   **Lowercasing and Punctuation Removal:** Standardize text format.

3.  **Alternative Feature Engineering:** Explore other text vectorization techniques:
    *   **TF-IDF (Term Frequency-Inverse Document Frequency):** This method weighs word importance, giving higher values to words that are rare but significant in a document and lower values to common words.
    *   **Word Embeddings (e.g., Word2Vec, GloVe):** Represent words as dense vectors in a continuous vector space, capturing semantic relationships between words.

4.  **More Complex Models:** Experiment with advanced machine learning and deep learning models:
    *   **Ensemble Methods:** Bagging, Boosting (e.g., RandomForest, Gradient Boosting).
    *   **Deep Learning Models:** Recurrent Neural Networks (RNNs), Long Short-Term Memory (LSTMs), or Convolutional Neural Networks (CNNs) for text.
    *   **Transformer-based Models:** Fine-tuning pre-trained models like BERT, RoBERTa, or DistilBERT for potentially state-of-the-art performance, though at higher computational cost.

5.  **Larger and More Diverse Datasets:** Training the model on a broader and more varied collection of SMS messages could improve its generalization capabilities and robustness to new types of spam.

6.  **Model Deployment:** Integrate the trained model into a production environment. This could involve:
    *   Building a lightweight REST API (e.g., using Flask or FastAPI) to expose the model's prediction functionality.
    *   Deploying the API to cloud platforms (e.g., Google Cloud Run, AWS Lambda, Azure Functions).
    *   Saving and loading the model using `joblib` or `pickle` for persistence.

7.  **Explainable AI (XAI):** Implement techniques to understand *why* the model makes certain predictions. Tools like LIME or SHAP can provide insights into word importance for classification, increasing trust and interpretability.

## 🔮 Future Improvements

To further enhance this SMS spam classification project and build upon the current success, several avenues for future work can be explored:

1.  **Hyperparameter Tuning:** Systematically optimize the parameters of `CountVectorizer` (e.g., `ngram_range`, `max_features`, `min_df`) and `MultinomialNB` (e.g., `alpha` for Laplace smoothing) using techniques like GridSearchCV or RandomizedSearchCV.

2.  **Advanced Text Preprocessing:** Incorporate more sophisticated text cleaning steps:
    *   **Stop Word Removal:** Eliminate common words (e.g., 'the', 'a', 'is') that often carry little semantic value.
    *   **Lemmatization/Stemming:** Reduce words to their root form (e.g., 'running' -> 'run') to minimize vocabulary size and treat variations as the same word.
    *   **Lowercasing and Punctuation Removal:** Standardize text format.

3.  **Alternative Feature Engineering:** Explore other text vectorization techniques:
    *   **TF-IDF (Term Frequency-Inverse Document Frequency):** This method weighs word importance, giving higher values to words that are rare but significant in a document and lower values to common words.
    *   **Word Embeddings (e.g., Word2Vec, GloVe):** Represent words as dense vectors in a continuous vector space, capturing semantic relationships between words.

4.  **More Complex Models:** Experiment with advanced machine learning and deep learning models:
    *   **Ensemble Methods:** Bagging, Boosting (e.g., RandomForest, Gradient Boosting).
    *   **Deep Learning Models:** Recurrent Neural Networks (RNNs), Long Short-Term Memory (LSTMs), or Convolutional Neural Networks (CNNs) for text.
    *   **Transformer-based Models:** Fine-tuning pre-trained models like BERT, RoBERTa, or DistilBERT for potentially state-of-the-art performance, though at higher computational cost.

5.  **Larger and More Diverse Datasets:** Training the model on a broader and more varied collection of SMS messages could improve its generalization capabilities and robustness to new types of spam.

6.  **Model Deployment:** Integrate the trained model into a production environment. This could involve:
    *   Building a lightweight REST API (e.g., using Flask or FastAPI) to expose the model's prediction functionality.
    *   Deploying the API to cloud platforms (e.g., Google Cloud Run, AWS Lambda, Azure Functions).
    *   Saving and loading the model using `joblib` or `pickle` for persistence.

7.  **Explainable AI (XAI):** Implement techniques to understand *why* the model makes certain predictions. Tools like LIME or SHAP can provide insights into word importance for classification, increasing trust and interpretability.

In [ ]:
# 📋 Generate requirements.txt
# This cell creates a `requirements.txt` file listing all the Python libraries and their versions required to run this project.
# This ensures reproducibility of the environment.

requirements_content = """
pandas
numpy
scikit-learn
"""

with open('requirements.txt', 'w') as f:
    f.write(requirements_content)

print("Generated requirements.txt")

Generated requirements.txt


## 🔍 Exploratory Data Analysis (EDA)

Exploratory Data Analysis (EDA) is crucial for understanding the characteristics of our dataset. We'll examine its structure, check for imbalances, and get a general sense of the data before diving into preprocessing and modeling.

## 📂 Data Loading

Loading the dataset is the first step to bring our raw information into the analysis environment. We read the SMS spam dataset from a CSV file into a pandas DataFrame, which provides a structured and convenient way to work with the data.

## 📚 Import Libraries

This section is dedicated to importing all the necessary Python libraries. We use `pandas` for efficient data handling and `numpy` for numerical operations, which are fundamental for almost any data science project.

## ⚙️ Project Workflow

The project follows a standard machine learning pipeline:

1.  **Import Libraries**: Essential tools for data manipulation and model building.
2.  **Data Loading**: Ingesting the SMS message dataset.
3.  **Exploratory Data Analysis (EDA)**: Understanding the dataset's structure and characteristics.
4.  **Data Cleaning & Preprocessing**: Cleaning and preparing text data for model input.
5.  **Feature Engineering**: Converting text into numerical features suitable for machine learning algorithms.
6.  **Model Building & Training**: Selecting and training a classification model.
7.  **Model Evaluation**: Assessing the model's performance using various metrics.
8.  **Prediction**: Demonstrating the model's ability to classify new messages.
9.  **Conclusion & Future Work**: Summarizing findings and outlining potential improvements.